### ISSUE TITLE
Support merging with an array item in composition patches

### ISSUE THREAD

**Issue by ulucinar** [crossplane/crossplane#3335]
### What problem are you facing?

We have recently encountered an issue with a Crossplane composition that patches the selector field of an array item. Crossplane [does a merge patch](https://github.com/crossplane/crossplane-runtime/blob/619edbacb81eb9b4ccb6a5a7ba8f07d74cc07dc9/pkg/resource/api.go#L157), which overrides the whole array element together with the target field and the reference field. This causes the already resolved target fields to get lost, leaving only the selector field. 

In such cases, there is a race condition between the Crossplane's composite reconciler (who composes and patches the composed resource) and the associated Crossplane provider's managed resource reconciler.

Here's how the situation can be observed:
```yaml
---
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.9.2
  creationTimestamp: null
  name: myresources.xp.test.com
spec:
  group: xp.test.com
  names:
    kind: MyResource
    listKind: MyResourceList
    plural: myresources
    singular: myresource
  scope: Cluster
  versions:
  - name: v1alpha1
    schema:
      openAPIV3Schema:
        description: MyResource is the Schema for the myresources API
        properties:
          apiVersion:
            description: 'APIVersion defines the versioned schema of this representation
              of an object. Servers should convert recognized schemas to the latest
              internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
            type: string
          kind:
            description: 'Kind is a string value representing the REST resource this
              object represents. Servers may infer this from the endpoint the client
              submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
            type: string
          metadata:
            type: object
          spec:
            description: MyResourceSpec defines the desired state of MyResource
            properties:
              forProvider:
                description: A is an example field of MyResource. Edit myresource_types.go
                  to remove/update
                properties:
                  b:
                    properties:
                      targetRefs:
                        items:
                          description: A Reference to a named object.
                          properties:
                            name:
                              description: Name of the referenced object.
                              type: string
                            policy:
                              description: Policies for referencing.
                              properties:
                                resolution:
                                  default: Required
                                  description: Resolution specifies whether resolution
                                    of this reference is required. The default is
                                    'Required', which means the reconcile will fail
                                    if the reference cannot be resolved. 'Optional'
                                    means this reference will be a no-op if it cannot
                                    be resolved.
                                  enum:
                                  - Required
                                  - Optional
                                  type: string
                                resolve:
                                  description: Resolve specifies when this reference
                                    should be resolved. The default is 'IfNotPresent',
                                    which will attempt to resolve the reference only
                                    when the corresponding field is not present. Use
                                    'Always' to resolve the reference on every reconcile.
                                  enum:
                                  - Always
                                  - IfNotPresent
                                  type: string
                              type: object
                          required:
                          - name
                          type: object
                        type: array
                      targetSelector:
                        description: A Selector selects an object.
                        properties:
                          matchControllerRef:
                            description: MatchControllerRef ensures an object with
                              the same controller reference as the selecting object
                              is selected.
                            type: boolean
                          matchLabels:
                            additionalProperties:
                              type: string
                            description: MatchLabels ensures an object with matching
                              labels is selected.
                            type: object
                          policy:
                            description: Policies for selection.
                            properties:
                              resolution:
                                default: Required
                                description: Resolution specifies whether resolution
                                  of this reference is required. The default is 'Required',
                                  which means the reconcile will fail if the reference
                                  cannot be resolved. 'Optional' means this reference
                                  will be a no-op if it cannot be resolved.
                                enum:
                                - Required
                                - Optional
                                type: string
                              resolve:
                                description: Resolve specifies when this reference
                                  should be resolved. The default is 'IfNotPresent',
                                  which will attempt to resolve the reference only
                                  when the corresponding field is not present. Use
                                  'Always' to resolve the reference on every reconcile.
                                enum:
                                - Always
                                - IfNotPresent
                                type: string
                            type: object
                        type: object
                      targets:
                        items:
                          type: string
                        type: array
                    type: object
                  bArr:
                    items:
                      properties:
                        targetRefs:
                          items:
                            description: A Reference to a named object.
                            properties:
                              name:
                                description: Name of the referenced object.
                                type: string
                              policy:
                                description: Policies for referencing.
                                properties:
                                  resolution:
                                    default: Required
                                    description: Resolution specifies whether resolution
                                      of this reference is required. The default is
                                      'Required', which means the reconcile will fail
                                      if the reference cannot be resolved. 'Optional'
                                      means this reference will be a no-op if it cannot
                                      be resolved.
                                    enum:
                                    - Required
                                    - Optional
                                    type: string
                                  resolve:
                                    description: Resolve specifies when this reference
                                      should be resolved. The default is 'IfNotPresent',
                                      which will attempt to resolve the reference
                                      only when the corresponding field is not present.
                                      Use 'Always' to resolve the reference on every
                                      reconcile.
                                    enum:
                                    - Always
                                    - IfNotPresent
                                    type: string
                                type: object
                            required:
                            - name
                            type: object
                          type: array
                        targetSelector:
                          description: A Selector selects an object.
                          properties:
                            matchControllerRef:
                              description: MatchControllerRef ensures an object with
                                the same controller reference as the selecting object
                                is selected.
                              type: boolean
                            matchLabels:
                              additionalProperties:
                                type: string
                              description: MatchLabels ensures an object with matching
                                labels is selected.
                              type: object
                            policy:
                              description: Policies for selection.
                              properties:
                                resolution:
                                  default: Required
                                  description: Resolution specifies whether resolution
                                    of this reference is required. The default is
                                    'Required', which means the reconcile will fail
                                    if the reference cannot be resolved. 'Optional'
                                    means this reference will be a no-op if it cannot
                                    be resolved.
                                  enum:
                                  - Required
                                  - Optional
                                  type: string
                                resolve:
                                  description: Resolve specifies when this reference
                                    should be resolved. The default is 'IfNotPresent',
                                    which will attempt to resolve the reference only
                                    when the corresponding field is not present. Use
                                    'Always' to resolve the reference on every reconcile.
                                  enum:
                                  - Always
                                  - IfNotPresent
                                  type: string
                              type: object
                          type: object
                        targets:
                          items:
                            type: string
                          type: array
                      type: object
                    type: array
                type: object
            type: object
          status:
            description: MyResourceStatus defines the observed state of MyResource
            type: object
        type: object
    served: true
    storage: true
    subresources:
      status: {}

---
apiVersion: apiextensions.crossplane.io/v1
kind: CompositeResourceDefinition
metadata:
  name: xmyresources.test.com
spec:
  group: test.com
  names:
    kind: XMyResource
    plural: xmyresources
  versions:
  - name: v1alpha1
    served: true
    referenceable: true
    schema:
      openAPIV3Schema:
        type: object
        properties:
          spec:
            type: object
            properties:
              name:
                type: string
            required:
              - name

---
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: example
  labels:
    purpose: example
spec:
  compositeTypeRef:
    apiVersion: test.com/v1alpha1
    kind: XMyResource
  resources:
    - name: myresource
      base:
        apiVersion: xp.test.com/v1alpha1
        kind: MyResource
        metadata:
        spec:
          forProvider: {}
      patches:
        - fromFieldPath: "spec.name"
          toFieldPath: spec.forProvider.bArr[0].targetSelector.matchLabels[key2]

---
apiVersion: test.com/v1alpha1
kind: XMyResource
metadata:
  name: test
spec:
  name: testName
  compositionRef:
    name: example
```

If, after applying these manifests, the following patch is applied on the composed resource (simulating the provider's action):
```yaml
spec:
  forProvider:
    b:
      targets:
      - testTarget1.a
      - testTarget1.b
      targetRefs:
      - name: targetName1.a
      - name: targetName1.b
      targetSelector:
        matchLabels:
          key1: val1
    bArr:
      - targets:
        - testTarget2.a
        - testTarget2.b
        targetRefs:
        - name: targetName2.a
        - name: targetName2.b
        targetSelector:
          matchLabels:
            key2: val2
```
, one can observe that the `spec.forProvider.bArr[0].targets` and `spec.forProvider.bArr[0].targetRefs` fields are emptied.

### How could Crossplane help solve your problem?

We can introduce a new patch policy that will merge, instead of replace, array items. The current behavior of [JSON merge patch](https://tools.ietf.org/html/rfc7386) works as expected but the resulting behavior is not intuitive and can easily cause issues.

**Comment by bobh66** [comment]
Did you try the mergeOptions appendSlice policy as described here:  https://crossplane.io/docs/v1.9/reference/composition.html#compositions ?

**Comment by ulucinar** [comment]
Hi @bobh66,
Thank you for the pointer. I think the appendSlice is for appending a new slice item in addition to the existing ones in the observed state when patching. However, what we need here is to be able to merge with an existing slice item via a designated key. This would be similar to Kubernetes strategic merge patching with a similar need for the specification of a `patchMergeKey` in addition to specifying the patch policy.

**Comment by negz** [comment]
Is there any upstream machinery that could help with this? In particular I suspect we're overdue to move to server-side apply which appears to have 'built in' support for merge strategies on custom resources - https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy

**Comment by muvaf** [comment]
If I'd like to patch `spec.forProvider.bArr[0].targetSelector.matchLabels[key2]` and other properties of `spec.forProvider.bArr[0]` get affected by this for some reason, I'd say this is a bug regardless of the underlying machinery. So I changed the label to `bug`.

**Comment by negz** [comment]
@muvaf I don't disagree - my point is that I'm wary of continuing to "reinvent the wheel" by teaching our client-side patching logic how to do what it seems SSA already does.

**Comment by muvaf** [comment]
@negz Oh yeah, my comment wasn't really directed towards what approach we'd like to take. Just wanted to change it to be considered as a bug, which I think would impact whether we add a policy if we consider it as feature or improve the existing behavior if it's considered as bug.

From what I'm reading the doc you shared, `Using Server-Side Apply in a controller` section does describe the behavior we'd like so I'd like to give it a try and if it works that'd be my preference as well.

If it doesn't, I think another option could be to generate a set of `fieldpath.SetValue` statements from the given template and prepend it to the statements produced from `patches` array of the composition. Then we'd fetch the existing object, run all `fieldpath.SetValue` statements on it and then call `Update` instead of patching. It's indeed a bit reinventing the wheel if the SSA would work out of box.

**Comment by muvaf** [comment]
I just gave it a shot. Added an applicator to crossplane-runtime in https://github.com/crossplane/crossplane-runtime/pull/358 that uses server-side apply as described [here](https://github.com/kubernetes-sigs/controller-runtime/issues/347).

First, I wanted to reproduce the behavior with what @ulucinar provided. Just needed to add the following to give Crossplane permissions to work with the CRD:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: add-crd-permissions
  labels:
    crossplane.io/scope: "system"
    rbac.crossplane.io/aggregate-to-crossplane: "true"
rules:
  - apiGroups:
      - xp.test.com
    resources:
      - myresources
    verbs:
      - '*'
```

Then I applied the claim and went ahead and `kubectl edit`ed the following managed resource:
```yaml
spec:
  forProvider:
    bArr:
    - targetSelector:
        matchLabels:
          key2: testName
```
to be like the following:
```yaml
spec:
  forProvider:
    bArr:
    - targetRefs:
      - name: ola 
      targetSelector:
        matchLabels:
          key2: testName
```

Then I confirmed that in the next reconciliation, the `targetRefs` field is gone. Then I shut down Crossplane pod and started my local Crossplane that uses the new server-side applicator, only needed to change [this line](https://github.com/crossplane/crossplane/blob/b0437b7/internal/controller/apiextensions/composite/reconciler.go#L320) after making it use my local crossplane-runtime from [this draft PR](https://github.com/crossplane/crossplane-runtime/pull/358).

Once it's back up, unfortunately, I got the same behavior. Please do check my code to see if I used SSA properly. If it's correct and we're sure it doesn't work for us, we can start working on a client patching approach implemented as another `Applicator`.

**Comment by muvaf** [comment]
I run another experiment today with the OpenAPI markers. In SSA (it doesn't work with client-side), you can designate a field to be the identifier of the array item and I did that by adding another field called `index` to the list above. The schema became the following for `bArr` field:
```yaml
bArr:
  items:
    properties:
      index:
        type: integer
      targetRefs: ...
      targetSelector: ...
      targets: ...
    type: object
    required: ["index"]
  type: array
  x-kubernetes-list-type: map
  x-kubernetes-list-map-keys: ["index"]
```

Then in the `Composition` template I assigned `0` to the `index` of the first element. It worked. When I added `targetRefs: [{name: "ola"}]` to the `0`th element of the array, it didn't get erased because SSA treated the list as a map by using `index` as key, hence instead of replacing the list it applied the patches of each element to their corresponding in the api-server. When I used our classic patching applicator, the additional properties weren't effective, they only worked with SSA.

What I'm wondering now is whether we can get away by having `x-kubernetes-list-map-keys` use the index of the item to identify it so that we don't have to add additional field to every array item struct. Just asked on [Kubernetes Slack](https://kubernetes.slack.com/archives/C0EG7JC6T/p1664442841067319).

**Comment by muvaf** [comment]
I've just updated the client-side experiment https://github.com/crossplane/crossplane/pull/3347 . Now, it's just a new `ApplyOption` that patches current object with every field of the desired state. It's working as expected with the test data @ulucinar provided.

**Comment by negz** [comment]
I think we should be biasing for moving toward SSA rather than extending (further diverging) our client-side patching logic.

> But if there isn't, it's a huge overhead to have to think about what to use as key for each array and add & maintain when there isn't one (like the given example, and keep in mind the keys should be scalar).

Could we establish a pattern where we by default inject a string field "`name`" into arrays of objects (as in "name of this entry in the array", and use that as a map key? I feel like there's quite a bit of precedent for this in Kubernetes - e.g. the `containers` array in a Pod, and also the `resources` array in our Compositions. I certainly regret that we didn't originally include `name` in Compositions in order to allow the array to be safely reordered (e.g. removing elements, injecting intermediate elements). We could override this key to point to a more sensible existing object field if/when one exists.

I think the key challenges with this approach are:

* It's a breaking change if we make the new field required.
* Anywhere we use `spec.array[0]` notation in a field path is already relying on stable array ordering.

**Comment by muvaf** [comment]
> I think we should be biasing for moving toward SSA rather than extending (further diverging) our client-side patching logic.

Agreed but I think there should be a ceiling to how much sacrifice we can make.

> It's a breaking change if we make the new field required.

The map keys need to be either required or defaulted. Defaulting would be fine and non-breaking but defeats the purpose of unique identification. So, it'd be breaking change. (I just checked that apiserver enforces that uniqueness of given keys of the elements in the array. So, I'm guessing that they allow defaulting to accommodate single-item arrays (?). After the single element, it practically becomes required without a default.)

> Anywhere we use `spec.array[0]` notation in a field path is already relying on stable array ordering.

I'm thinking about this in a way that maybe we can have a map-key based selector instead of index but ending up finding myself coming up with a new Crossplane-specific format that's surprising when your expectation is just a JSON field path. And that is on top of the overhead of adding an identifier to each array item struct.

I see the following format in `jq` for example:
```
jq '.DATA[]  | select(.DOMAIN == "domain2") | .DOMAINID'
```

What's close to that could be like the following:
```
spec.array[name==myitemname]
```

Once you get there, then we could be looking at expanding it to more than a single field, which is what SSA allows you to do, i.e. `spec.array[someProperty==ola,anotherProperty==oma]`.

So it got me thinking that with these selection mechanisms, we're essentially treating all arrays as a set and trying not to patch the wrong object. But from the eyes of the API consumer, is it ever treated as a set? i.e. whatever we do, it's essentially a list with an order for user if the API designer didn't make it a map. If the user will always see and treat it as an ordered list anyway, then that feels like we're trying to solve a problem that doesn't exist for the user, i.e. _what if a list is actually a set_. It goes back to Kubernetes [API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps) and I went down the rabbit hole of the discussions to understand the main reason of the convention. The [slack conversation](https://kubernetes.slack.com/archives/C0EG7JC6T/p1664442841067319) turned into a nice discussion as well.

Coming back, I see the following arguments for two options.
* Use a property as key, add a new property if it doesn't exist.
  * Unless we come up with a special field path selector logic, this doesn't remove our reliance on the order of elements, hence ends up adding yet another thing to think about.
  * Breaks high fidelity principle on managed resources. If the external API designed it to be a list, the order may indeed be important.
  * A plus is that late initialization would be easier when the defaults of the external API includes adding an item. It's rare but when that happens, we are forced to come up with a differentiator string which may include the total of all properties. (SecurityGroup rules used to be an example but we separated them out into their own CRD, which was what should've been done in the first place)
* Use Update API with client side patching.
  * You need to fetch the existing object (this is a good thing in some respects since it guarantees you're working on the most recent resourceVersion)
  * `Create if not exists` logic needs to be implemented in the client.
  * Need to maintain our client-side patch implementation.
  * Unless we come up with a property-based selector for arrays, it's not possible to choose an item in the array via a property other than index when you use composition patches.

The key points for me are:
* It's too surprising for folks to have a patch with field path `spec.arr[0].property` have an impact on `spec.arr[0].anotherProperty`. 
* We cannot prevent usage of indexes in field patches of composition, i.e. `spec.arr[0]`, unless we implement our own selector syntax, take on the big challenge of migrating people off of it and prevent it programmatically in composition and come up with a linter that requires every array in CRDs to have a map key. 
* It is surprising for folks to see a required field under `spec.forProvider` that is not part of the external API.

The ideal state I'd like us to be in is one where there is a standard syntax for property-based selector in field path and regardless of the schema of the managed resource, users would be able to reference an item by a property or an index - this is without imposing a requirement that the item struct would have a map key. Because I agree with api machinery folks that the property-based selection is important and we should have a way to do it but it doesn't have to be through SSA because we control both parties, i.e. patcher controller and patched resources. Because we control the client, we could replicate the behavior without enforcing a breaking change on the CRDs which is different than the assumptions of k8s apimachinery folks.

With that in mind, I think we should switch to Update API with client-side patch implementation I experimented with [here](https://github.com/crossplane/crossplane-runtime/pull/359) and then start a discussion around how we can add property-based selector to the field path syntax. We could also use that discussion as a proxy to see how many use cases out there would benefit from it. Once we decide to do it, we can discuss whether we should switch to SSA or do it in the client side in its design doc. At that point, the benefit of making CRDs compatible with SSA would be for user's interaction with the managed resource CRs rather than how composite reconciler works.

**Comment by muvaf** [comment]
There are good arguments here https://github.com/kubernetes/website/pull/37110 in favor of SSA even if we don't have map keys, like "extract" method to filter only the changed fields. We could possibly move to SSA without solving the problem this issue describes to enable provider developers use map keys if they choose to do so. Ideally, it'd be no-op for others, hence we can continue the discussion about whether we should introduce map keys or use index in the client to do the patches since SSA is compatible with both.

**Comment by negz** [comment]
Thanks for digging deep into this @muvaf.

I continue to feel that our ultimate goal should be to migrate to SSA. I can see that it might be unrealistic to block this particular fix on that migration, but I am a little wary of kicking the can down the road and never getting around to it.

Using "map keys" in an array of objects does seem more resilient than an ordered array due to the issues you discussed with @lavalamp around different actors updating a list. For lists of objects it does seem ideal to have "extra context" (i.e. a key field or fields) to help patching logic make better decisions, whether that's Composition patching logic mutating an object by field path, or SSA patching logic. To that end, I agree that it makes sense for us to expand field paths to support key-based lookups (e.g. `foo[name==bar]`).

I'm a little reluctant for Crossplane field paths to diverge from Kubernetes's very basic implementation, but that ship sailed when we added wildcard support.  I don't think we would need to deprecate and block index (e.g. `foo[0]`) style field paths - I would feel comfortable introducing a better way and updating our documentation to encourage folks to use the better approach. Anyone continuing to use index style field paths would be subject to the same issue they're subject to today - the element they intended to reference moving to another index.

I'm not as worried about injecting Crossplane-specific fields under `forProvider`, as long as they're standard and well documented. There's precedent for this (refs and selectors). It's not ideal, but I could live with it.

I noticed recently that Kyverno uses https://jmespath.org. This appears to be a (much) more advanced version of our field paths. It includes:

* The same base syntax as our field paths - e.g. `foo[0].name`.
* Wildcard expressions (e.g. `foo[*].name`)
* Filter expressions (e.g. `foo[?name==bar].value`)

It does however also include a [whole bunch of other features](https://jmespath.org/tutorial.html#pipe-expressions) that I'd be wary of adopting without good reason.

**Comment by lavalamp** [comment]
For reference, the serverside apply code serializing paths is here: https://github.com/kubernetes-sigs/structured-merge-diff/tree/master/fieldpath

e.g. this test illustrates multiple list keys: https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/fieldpath/path_test.go#L44

...I don't think this code is going to solve your problems (no wildcards, at least not intentionally--it might actually work though I guess), but it may be of some interest.

**Comment by negz** [comment]
@lavalamp Thanks, that helps - at least being able to align on the same format. We're currently using a similar package - https://github.com/crossplane/crossplane-runtime/tree/master/pkg/fieldpath. We'd probably have used the one you linked rather than reinventing the wheel but we weren't aware of its existence.

**Comment by muvaf** [comment]
Seems like extract methods aren't available through controller-runtime. I need to spend more time on this to make sure we know precisely whether it's possible at all to have index-based patching in SSA. I'd be more comfortable if we know all implications of SSA before making a decision either way.

The fix for this issue won't make it in `v1.10.0` whose code freeze date is today.

**Comment by apelisse** [comment]
> Seems like extract methods aren't available through controller-runtime

Side-note, can you open a bug against controller-runtime for that, so that we can keep track of it.

**Comment by jbw976** [comment]
There was a fair amount of investigation into how to approach this, and anecdotally it's bitten a number of people, I would definitely like to keep this one open.

**Comment by jbw976** [comment]
/fresh this is being tracked as part of our #4047 server side apply roadmap epic 💪

**Comment by ytsarev** [comment]
/fresh the issue is still valid, see fresh #5055

**Comment by ulucinar** [comment]
We've rolled out SSA patching while updating the resolved `subnetId` references in the `spec.forProvider.vpcConfig[]` object list in order to address https://github.com/upbound/provider-aws/issues/975 with [`upbound/provider-aws@v0.47.0`](https://github.com/upbound/provider-aws/releases/tag/v0.47.0). Also validated the fix with the [P&T composition function v0.2.1](https://marketplace.upbound.io/functions/upbound/function-patch-and-transform/v0.2.1) with the composition described [here](https://github.com/upbound/provider-aws/pull/1046#issue-2057334633).
